### ISSUE TITLE
WAM Table mismatch

### ISSUE THREAD

**Issue by md5i** [SWI-Prolog/swipl-devel#917]
Between Dec 18 and now, one particular machine is failing from a clean build with the following error:

```
[ 95%] Generating ../home/boot.prc
[FATAL ERROR: at Mon Dec 20 11:26:45 2021
	WAM Table mismatch: wam_table[8(h_mpz)] == wam_table[9(h_mpq)]
]
make[2]: *** [src/CMakeFiles/bootfile.dir/build.make:98: home/boot.prc] Aborted (core dumped)
```

I recreated this on my home machine, but, when bisecting to figure out which commit caused this, the error went away in a subsequent build.  The error has happened in two successive builds on the machine where I originally noticed the problem.

**Comment by friguzzi** [comment]
I encountered the same error, I'm on a Ubuntu 18.04.6 LTS (GNU/Linux 4.15.0-163-generic x86_64)
It happened after the latest apt-get upgrade performed yesterday

**Comment by josd** [comment]
Same here
```
Submodule path 'packages/swipl-win': checked out 'c0ed8afaf78fa2a51ee7391ea12536b092c31300'
-- Configuring SWI-Prolog-8.5.4
-- Could NOT find LibTCMalloc (missing: LIBTCMALLOC_LIBRARY)
-- Could NOT find LibTCMalloc (missing: LIBTCMALLOC_LIBRARY)                               To activate the Python tests: python3 -mpip install protobuf                               -- Could NOT find JNI (missing: JAVA_AWT_LIBRARY JAVA_INCLUDE_PATH JAVA_INCLUDE_PATH2 JAVA_AWT_INCLUDE_PATH)                                                                          -- Could NOT find Qt5Widgets (missing: Qt5Widgets_DIR)                                     -- Configuring done                                                                        -- Generating done                                                                         -- Build files have been written to: /home/jdroo/github.com/SWI-Prolog/swipl-devel/build   [151/253] Generating ../home/boot.prc
FAILED: home/boot.prc                                                                      cd /home/jdroo/github.com/SWI-Prolog/swipl-devel/build/src && /usr/bin/cmake -E remove -f /home/jdroo/github.com/SWI-Prolog/swipl-devel/build/home/boot.prc && /home/jdroo/github.com/SWI-Prolog/swipl-devel/build/src/swipl -q -O -o /home/jdroo/github.com/SWI-Prolog/swipl-devel/build/home/boot.prc -b /home/jdroo/github.com/SWI-Prolog/swipl-devel/build/home/boot/init.pl                                                                                       [FATAL ERROR: at Tue Dec 21 00:29:45 2021                                                          WAM Table mismatch: wam_table[8(h_mpz)] == wam_table[9(h_mpq)]                     ]                                                                                          Aborted
[160/253] Building C object packages/ssl/CMakeFiles/plugin_crypto4pl.dir/crypto4pl.c.o     ninja: build stopped: subcommand failed.                                                   [8/95] Generating ../home/boot.prc                                                         FAILED: home/boot.prc                                                                      cd /home/jdroo/github.com/SWI-Prolog/swipl-devel/build/src && /usr/bin/cmake -E remove -f /home/jdroo/github.com/SWI-Prolog/swipl-devel/build/home/boot.prc && /home/jdroo/github.com/SWI-Prolog/swipl-devel/build/src/swipl -q -O -o /home/jdroo/github.com/SWI-Prolog/swipl-devel/build/home/boot.prc -b /home/jdroo/github.com/SWI-Prolog/swipl-devel/build/home/boot/init.pl                                                                                       [FATAL ERROR: at Tue Dec 21 00:30:03 2021                                                          WAM Table mismatch: wam_table[8(h_mpz)] == wam_table[9(h_mpq)]                     ]                                                                                          Aborted                                                                                    [17/95] Building C object packages/semweb/CMakeFiles/plugin_rdf_db.dir/rdf_db.c.o          ninja: build stopped: subcommand failed.
```

**Comment by josd** [comment]
Same here:

Submodule path 'packages/swipl-win': checked out 'c0ed8afaf78fa2a51ee7391ea12536b092c31300'
-- Configuring SWI-Prolog-8.5.4
-- Could NOT find LibTCMalloc (missing: LIBTCMALLOC_LIBRARY)
-- Could NOT find LibTCMalloc (missing: LIBTCMALLOC_LIBRARY)                               To activate the Python tests: python3 -mpip install protobuf                               -- Could NOT find JNI (missing: JAVA_AWT_LIBRARY JAVA_INCLUDE_PATH JAVA_INCLUDE_PATH2 JAVA_AWT_INCLUDE_PATH)                                                                          -- Could NOT find Qt5Widgets (missing: Qt5Widgets_DIR)                                     -- Configuring done                                                                        -- Generating done                                                                         -- Build files have been written to: /home/jdroo/github.com/SWI-Prolog/swipl-devel/build   [151/253] Generating ../home/boot.prc
FAILED: home/boot.prc                                                                      cd /home/jdroo/github.com/SWI-Prolog/swipl-devel/build/src && /usr/bin/cmake -E remove -f /home/jdroo/github.com/SWI-Prolog/swipl-devel/build/home/boot.prc && /home/jdroo/github.com/SWI-Prolog/swipl-devel/build/src/swipl -q -O -o /home/jdroo/github.com/SWI-Prolog/swipl-devel/build/home/boot.prc -b /home/jdroo/github.com/SWI-Prolog/swipl-devel/build/home/boot/init.pl                                                                                       [FATAL ERROR: at Tue Dec 21 00:29:45 2021                                                          WAM Table mismatch: wam_table[8(h_mpz)] == wam_table[9(h_mpq)]                     ]                                                                                          Aborted
[160/253] Building C object packages/ssl/CMakeFiles/plugin_crypto4pl.dir/crypto4pl.c.o     ninja: build stopped: subcommand failed.                                                   [8/95] Generating ../home/boot.prc                                                         FAILED: home/boot.prc                                                                      cd /home/jdroo/github.com/SWI-Prolog/swipl-devel/build/src && /usr/bin/cmake -E remove -f /home/jdroo/github.com/SWI-Prolog/swipl-devel/build/home/boot.prc && /home/jdroo/github.com/SWI-Prolog/swipl-devel/build/src/swipl -q -O -o /home/jdroo/github.com/SWI-Prolog/swipl-devel/build/home/boot.prc -b /home/jdroo/github.com/SWI-Prolog/swipl-devel/build/home/boot/init.pl                                                                                       [FATAL ERROR: at Tue Dec 21 00:30:03 2021                                                          WAM Table mismatch: wam_table[8(h_mpz)] == wam_table[9(h_mpq)]                     ]                                                                                          Aborted

**Comment by JanWielemaker** [comment]
This should already be fixed with 7bf463db4dcc7b4e27c820ef8fc6f94fd8116720, the latest git patch pushed Dec 20 10:05:52.  Could you verify you do (not) have this patch and whether it works?

**Comment by friguzzi** [comment]
I have that patch and I still get the error

**Comment by JanWielemaker** [comment]
I see.   Yip, the compiler can still outsmart us.   Pushed 8f3b66d949e6c58d311cd5ccf774fcdbec9997a4 to (hopefully) fixes this.  I understand the compiler could squash these two instructions.   I do not understand why the compiler does this what seems rather randomly.   This is (I think) just high level optimization which one would expect to only depend on the compiler version.

Anyway, please try and report/close.

**Comment by friguzzi** [comment]
It is now working for me

**Comment by josd** [comment]
It is now also working for me.
The problem occured on a machine with gcc version 7.5.0 on Ubuntu 18.04.6 .
It did not occur on my other machine with gcc version 9.3.0 on Ubuntu 20.04.3 .

**Comment by JanWielemaker** [comment]
Ok.   Time to close :smile:

**Comment by md5i** [comment]
I'd like to re-open this issue, as it is still happening for me.  Using gcc 11.2.0 on SWIPL revno 8f3b66d949e6c58d311cd5ccf774fcdbec9997a4.

```
[ 95%] Generating ../home/boot.prc
[FATAL ERROR: at Tue Dec 21 12:32:49 2021
        WAM Table mismatch: wam_table[119(s_static)] == wam_table[124(s_multifile)]
]
Aborted
make[2]: *** [src/CMakeFiles/bootfile.dir/build.make:98: home/boot.prc] Error 134
```

**Comment by JanWielemaker** [comment]
Luckily that was an easy one.   The original was about the `SEPARATE_VMI` macro not working.   This is about a having forgotten one.   Added (and keeping closed :smile:  )   I hope these are all.  We now know what to do if one emerges.

**Comment by josd** [comment]
While retesting with gcc 7.5.0 I now got
```
[FATAL ERROR: at Wed Dec 22 16:12:31 2021
        WAM Table mismatch: wam_table[120(s_dynamic)] == wam_table[124(s_multifile)]
]
```

**Comment by JanWielemaker** [comment]
:cry:  Fixed with 9d2babb2c07e03bd57a03a596ec0d61fbe0b0635.   Overlooked as I thought the VMIs where different, but actually `enterDefinition(DEF)` has become a no-op, making the two VMIs equal ...   I still wonder why some people get these and I've never seen that.   Are you using specific CFLAGS?

**Comment by josd** [comment]
Now it works fine, thanks Jan!

We don't use specific CFLAGS and just run the following:
```
export CTEST_OUTPUT_ON_FAILURE=y
pushd ~/github.com/SWI-Prolog/swipl-devel
git pull
git submodule update
mkdir -p build
cd build
cmake -G Ninja ..
ninja
ctest -j 8
sudo ninja install 
```
